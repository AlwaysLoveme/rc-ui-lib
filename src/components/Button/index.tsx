import React, { useMemo, useCallback } from 'react';
import classnames from 'classnames';
import { ButtonProps } from './Props';
import Loading from '../Loading';
import Icon from '../Icon';
import styles from './index.scss';

const baseClass = 'r-button';
const Button: React.FC<ButtonProps> = (props) => {
    const {
        type = 'default',
        plain = false,
        disabled = false,
        loading = false,
        round = false,
        square = false,
        block = false,
        hairline = false,
        size = 'normal',
        loadingType = 'circular',
        loadingText,
        loadingSize = '16px',
        text,
        color,
        className,
        tag = 'button',
        nativeType,
        onClick,
        // touchstart,
        icon,
        iconPrefix = 'r-icon',
        iconPosition = 'left',
        children
    } = props;
    const Tag: any = tag || 'button';
    const Cls = classnames(
        styles[`${baseClass}`],
        {
            [styles[`${baseClass}--${type}`]]: !!type,
            [styles[`${baseClass}--${size}`]]: !!size,
            [styles[`${baseClass}--plain`]]: plain,
            [styles[`${baseClass}--disabled`]]: disabled,
            [styles[`${baseClass}--loading`]]: loading,
            [styles[`${baseClass}--round`]]: round,
            [styles[`${baseClass}--square`]]: square,
            [styles[`${baseClass}--block`]]: block,
            [styles[`${baseClass}--hairline`]]: hairline,
            [styles[`${baseClass}__${icon}`]]: !!icon
        },
        className
    );

    const renderIcon = useMemo(() => {
        if (loading) {
            return <Loading color={color} size={loadingSize} type={loadingType} />;
        }
        if (icon) {
            return (
                <Icon
                    className={classnames(styles[`${baseClass}__icon`])}
                    classPrefix={iconPrefix}
                    color={color}
                    name={icon}
                />
            );
        }
    }, [loading, icon, loadingType, loadingSize, color, iconPrefix]);

    const renderContent = useMemo(() => {
        let content;
        if (loading) {
            content = loadingText;
        } else {
            content = children || text;
        }
        if (content) {
            return <span className={classnames(styles[`${baseClass}__text`])}>{content}</span>;
        }
    }, [loading, loadingText, text, children]);

    const getStyle = useCallback(() => {
        if (color) {
            const style = {} as any;
            style.color = plain ? color : 'white';
            if (!plain) {
                // 为了使linear-gradient起作用，请用background 代替 backgroundColor
                style.background = color;
                style.border = 'none';
            }
            // 当是渐变色是，去掉边框。
            if (color.indexOf('gradient') !== -1) {
                style.border = 'none';
            } else {
                style.borderColor = color;
            }
            return style;
        }
        return null;
    }, [color, plain]);

    const handleClick = (e: TouchEvent | MouseEvent) => {
        if (!disabled || !loading) {
            onClick?.(e);
        }
    };

    return (
        <Tag
            className={Cls}
            disabled={disabled}
            onClick={handleClick}
            style={getStyle()}
            type={nativeType}
        >
            <div className={classnames(styles[`${baseClass}__content`])}>
                {iconPosition === 'left' && renderIcon}
                {/* {renderContent()} */}
                {renderContent}
                {iconPosition === 'right' && renderIcon}
            </div>
        </Tag>
    );
};

export default Button;
