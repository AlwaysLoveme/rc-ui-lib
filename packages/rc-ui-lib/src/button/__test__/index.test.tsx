import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from '../index';


describe('Button', () => {
  
it('should emit click event', () => {
  const onClick = jest.fn();
  const wrapper = mount(<Button onClick={onClick}>Click</Button>);
  wrapper.simulate('click');
  expect(onClick).toBeCalled();
});

it('should not emit click event when disabled', () => {
  const onClick = jest.fn();
  const wrapper = mount(<Button disabled onClick={onClick}>disabled</Button>);
  wrapper.simulate('click');
  expect(onClick).not.toBeCalled();
});

it('should not emit click event when loading', () => {
  const onClick = jest.fn();
  const wrapper = mount(<Button loading onClick={onClick}>loading</Button>);
  wrapper.simulate('click');
  expect(onClick).not.toBeCalled();
});

it('should hide border when color is gradient', () => {
  const wrapper = mount(
    <Button color="linear-gradient(#000, #fff)" />
  )
  expect(wrapper.getDOMNode().style.border).toEqual('0px');
});

it('should change icon class prefix when using icon-prefix prop', () => {
  const wrapper = mount(
    <Button icon="success" iconPrefix="my-icon" />
  )
  expect(toJson(wrapper.html())).toMatchSnapshot();
});

it('should render loading of a specific size when using loading-size prop', () => {
  const wrapper = mount(
    <Button loading loadingSize="10px" />
  ) 
  const loading = wrapper.find('.rc-loading__spinner');
  expect(loading.getDOMNode().style.width).toEqual('10px');
  expect(loading.getDOMNode().style.height).toEqual('10px');
});

it('should render icon in the right side when setting icon-position to right', () => {
  const wrapper = mount(
    <Button icon="plus" iconPosition="right" />
  ) 
  expect(toJson(wrapper)).toMatchSnapshot();
});


});