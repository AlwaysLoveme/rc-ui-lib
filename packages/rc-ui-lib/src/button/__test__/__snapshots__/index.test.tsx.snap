// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<button
  class="rc-button rc-button--default rc-button--normal"
>
  <div
    class="rc-button__content"
  />
</button>
`;

exports[`Button should change icon class prefix when using icon-prefix prop 1`] = `undefined`;

exports[`Button should render empty button without errors 1`] = `
<Button
  iconPosition="left"
  size="normal"
  tag="button"
  type="default"
>
  <button
    className="rc-button rc-button--default rc-button--normal"
    onClick={[Function]}
    style={Object {}}
  >
    <div
      className="rc-button__content"
    >
      <span
        className="rc-button__text"
        key="text"
      />
    </div>
  </button>
</Button>
`;

exports[`Button should render icon in the right side when setting icon-position to right 1`] = `
<Button
  icon="plus"
  iconPosition="right"
  size="normal"
  tag="button"
  type="default"
>
  <button
    className="rc-button rc-button--default rc-button--normal"
    onClick={[Function]}
    style={Object {}}
  >
    <div
      className="rc-button__content"
    >
      <Icon
        className="rc-button__icon"
        classPrefix="van-icon"
        name="plus"
        tag="i"
      >
        <Badge
          className="rc-button__icon van-icon van-icon-plus"
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper rc-button__icon van-icon van-icon-plus"
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
  </button>
</Button>
`;

exports[`Button should render icon slot correctly 1`] = `undefined`;
